standalone-demo/rating-element.js

```diff
@@ -1,71 +1,100 @@
-    this.setAttribute('value', val);
-    this._value = val;
-    this.highlight(val - 1);
+
+  connectedCallback() {
+    this.addEventListener('mouseleave', this.onHighlight);
+    this.addEventListener('mouseover', this.onHighlight);
+    this.addEventListener('click', this.onRate);
+
+    if (!this.hasAttribute('value')) this.setAttribute('value', 0);
+    if (!this.hasAttribute('max')) this.setAttribute('max', 5);
   }
 
-  get number() {
-    return this._number;
+  disconnectedCallback() {
+    this.removeEventListener('mouseleave', this.onHighlight);
+    this.removeEventListener('mouseover', this.onHighlight);
+    this.removeEventListener('click', this.onRate);
   }
-  set number(num) {
-    this.setAttribute('number', num);
-    this._number = num;
-    this.createRatingElements();
+
+  attributeChangedCallback(name) {
+    if (name === 'max') this.renderChoices();
+    const predicate = (_, i) => i < this.value;
+    return toggleClass(this.choices, this._classes.checked, predicate);
   }
 
-  highlight(index) {
-    // Highlight all the ratings up to and including the index.
-    this.ratingElements.forEach((el, i) => {
-      el.classList.toggle(this._markedClass, i <= index);
+  renderChoices() {
+    this.choices.forEach(el => el.remove());
+    this.choices = Array.from({ length: this.max }, () => {
+      const el = document.createElement('div');
+      el.className = this._classes.choice;
+      return this.appendChild(el);
     });
   }
 
-  createRatingElements() {
-    this.clearRatingElements();
+  highlightChoices(end = -1) {
+    const predicate = (_, i) => i <= end;
+    return toggleClass(this.choices, this._classes.highlighted, predicate);
+  }
 
-    for (let i = 0; i < this.number; i++) {
-      const el = document.createElement('div');
-      el.className = this._defaultClass;
-      this.appendChild(el);
-      this.ratingElements.push(el);
-      el.addEventListener('mousemove', e => this.highlight(i));
-      el.addEventListener('click', e => {
-        this.value = i + 1;
-        this.dispatchEvent(new Event('rate'));
-      });
-    }
+  get value() {
+    return parseInt(this.getAttribute('value'), 10);
+  }
+
+  set value(val) {
+    val = parseInt(val, 10);
+    if (val < 0 || val > this.max) return;
+    this.setAttribute('value', val);
   }
 
-  clearRatingElements() {
-    this.ratingElements = [];
-    while (this.firstChild) {
-      this.removeChild(this.firstChild);
-    }
+  get max() {
+    return parseInt(this.getAttribute('max'), 10);
+  }
+
+  set max(val) {
+    val = parseInt(val, 10);
+    if (val < Math.max(this.value, 1)) return;
+    this.setAttribute('max', val);
   }
 }
 
-// Register element so browser knows what to do with it.
-window.customElements.define('x-rating-element', RatingElement);
+if (window.customElements && window.customElements.define) {
+  window.customElements.define('x-rating-element', RatingElement);
+}
+
+function toggleClass(elements, className, callback) {
```

Comment:
Why not part of the class to make it isolated?

Reply:
Well it doesn't work with `this` which means it is not class/domain specific and if that's true we have one more reason to move it away to emphasize that 😄 I do it mostly to send message to the reader...

Reply:
Btw it is now pure function which means it is immediately testable without mocks/stubs needed 😉